### ISSUE TITLE
A Unified Model for Multi-class Anomaly Detection (2022)

### ISSUE THREAD

**Issue by YoojLee** [YoojLee/paper_review#33]
![image](https://user-images.githubusercontent.com/52986798/234000494-2bba2135-5720-4b32-a096-551c0c4a5bce.png)

# Abstract
기존의 anomaly detection 방법론들이one-class-one-model scheme임을 지적하며, 여러 클래스에 대한 anomaly detection을 하나의 모델로 수행할 수 있는 unified framework를 제안함. 
특히, 기존 anomaly detection에서 많이 활용된 reconstruction networks의 경우에는 normal뿐 아니라 anomalous sample 역시 잘 복구하게 되는(이렇게 되는 경우, outlier를 찾아내기가 어려워짐) "identical shortcut" 문제가 발생하게 되는데, UniAD는 아래의 3가지를 통해 해당 문제를 해결하고자 함.

- Layer-wise Query Decoder
- Neighbor-masked attention
- Feature Jittering (정확히는 Feature Jittering은 identical shortcut 문제를 해결하기 위함보다는 denoising task를 주게 됨으로써 입력 이미지 복구를 좀 더 키우는 쪽에 가까울 듯)

MVTec-AD와 CIFAR-10 데이터셋에서 실험을 진행한 결과, 기존의 SOTA 알고리즘을 훨씬 뛰어넘는 결과를 보여줌 (MVTec-AD의 경우, detection과 localization에서 대략 8%, 7% 정도 앞지름)

# Motivation Summary
기존의 anomaly detection 방법론들은 각 object 별로 따로 모델을 학습시킴. 이러한 'one-class-one-model' scheme은 메모리 사용량을 증가시킨다는 단점이 있음. 특히, 이는 클래스의 개수가 증가할 때 더욱 문제가 될 수 있음. normal sample이 intra-class diversity가 매우 커질 경우에도 one-class-one-model scheme이 좋은 접근이라고 할 수 없음. (multi-class에서는 아무래도 multi-class normal에 대한 분포를 modeling하는 걸 목적으로 삼고 있으니 그런 부분에서 multi-class one-model scheme이 diversity가 큰 경우를 다룰 때에 간편한 방식이라고 볼 수 있을 듯하다)

논문에서는 이 'one-class-one-model' scheme의 비효율적임을 문제로 삼고, **multi-class anomaly detection**에 대한 필요성을 제기함과 동시에 이를 위한 방법론을 제안하고 있음. 제안한 방법론은 학습 후 추가적인 fine-tuning 없이도 여러 클래스에 대한 anomaly detection이 가능하다고 말하고 있음. 이때 클래스 정보가 전혀 주어지지 않았음(데이터 preparation에 있어서 더 간편)에도 높은 성능을 보여주고 있음.

아래는 multi-class anomaly detection setting에 대한 그림.

![image](https://user-images.githubusercontent.com/52986798/234001368-5de83ec7-11b8-4379-bf48-afe8c466c28c.png)

AD 방법론 중 popular approach인 reconstruction 기반의 방법론의 경우에 multi-class에 대한 정상 데이터 분포 학습 시 **identical shortcut**에 빠지는 문제가 발생함. 논문에서는 이러한 identical shortcut 문제를 transformer의 query embedding 내에서 해결의 실마리를 찾아냄. 이에 layer-wise query embedding과 neighbor-masked attention을 통해 해결함. 

아래는 기존의 mlp와 cnn 구조가 identical shortcut에 취약함을 보여주면서 UniAD는 이를 해결했음을 보여주는 그림.

![image](https://user-images.githubusercontent.com/52986798/234001472-7255aad4-3371-4990-a2e8-520f47d897f8.png)

# Method Summary

기존 fully connected layer와 convolution의 경우에는 수식 구조 상으로 weight가 identity matrix로 regress될 확률이 높다는 것을 밝히고 있음. FC layer를 예로 들면,

$$ y = x^+w + b \in \mathbb{R}^{K \times C} .$$

대표적인 reconstruction loss 중 하나인 MSE loss의 경우, 학습 과정 중 y를 x^+로 push하게 됨. 따라서, $w \rightarrow I, b \rightarrow 0$. 이렇게 될 경우, anomaly image $x^-$ 역시 reconstruct가 잘 된다는 문제가 발생함.

$1 \times 1$ convolution의 경우, fully connected와 동일하다는 점에서 역시 identical shortcut 문제에 취약함 (커널 사이즈가 1보다 더 크다고 하더라도 커널 사이즈가 1인 경우가 다루는 경우를 동일하게 수행할 수 있고, parameter와 capacity가 더 크다는 점에서 동일 문제에서 자유롭지 못함).

반면, transformer의 attention layer 경우에는 아래와 같이 수식이 정의됨.

$$y=\mathrm{softmax}(q(x^+)^T / \sqrt{C})x^+ \in \mathbb{R}^{K\times C}.$$

$y$를 $x^+$에 가깝게 하기 위해, attention map이 $I$를 approx해야 함. 따라서, $q$는 $x^+$와 매우 큰 연관을 가져야 함. 이러한 점에서 $q$는 $x^-$를 잘 복원하지 못할 것이라 가정 가능함.

논문에서는 이러한 점에서 착안하여 디코더 내 레이어 별로 쿼리 임베딩 $q$를 추가하고, 추가적인 cheating을 방지하기 위해서 attention layer에서 neighbor masking을 적용함.

![image](https://user-images.githubusercontent.com/52986798/234005083-10b2054d-043b-470f-ad22-6e020e38df10.png)

![image](https://user-images.githubusercontent.com/52986798/234005111-f0d815f2-3ccc-46fe-8e83-1a642c38b40a.png)

추가적으로 feature jittering을 적용하여, denoising task를 수행하도록 하여 reconstruction을 보다 잘할 수 있도록 함.

아래는 feature jittering의 노이즈 $D$가 샘플링되는 분포를 표현한 것임. $\alpha$는 noisy degree를 control하는 jittering scale.
![image](https://user-images.githubusercontent.com/52986798/234005348-8c5cd990-e37d-4283-bec5-17f7079303a6.png)

objective는 MSE Loss를 사용하고 있으며, feature reconstruction score map $s=||f_{org}-f_{rec}||_2 \in \mathbb{R}^{H \times W}$를 얻어내어 이를 bi-linear interpolation을 통해 anomaly localization results를 얻어냄. Image level의 anomaly detection results 같은 경우에는 $s$를 average pooling한 후, 최댓값을 taking함.

# Experiment highlights

## Quantitative results
![image](https://user-images.githubusercontent.com/52986798/234006358-bdccb635-899b-4a01-86ba-c047d3519060.png)
![image](https://user-images.githubusercontent.com/52986798/234006408-309ae0ec-100e-47e7-b583-19b957b09f70.png)
baseline의 경우, unified case에서 성능 하락이 매우 큼. 그러나, UniAD는 unified와 separate case 간의 성능 차이가 별로 없고, 실제로 unified case로 학습하는 경우 성능이 더 올라가기도 함 (localization의 경우). 

![image](https://user-images.githubusercontent.com/52986798/234007086-e99bf503-ac7f-4e9e-a631-59759d34c471.png)
CIFAR-10에서도 수행함. 이 경우에는 unified case를 normal 클래스 여럿 abnormal 클래스 여럿 이런 식으로 해서 태스크 세팅. 역시 baseline 대비 성능이 매우 우수함.

## Qualitative results
![image](https://user-images.githubusercontent.com/52986798/234006467-efe88626-606b-47f4-b492-ce87f4242a06.png)
anomaly의 경우 reconstruction이 제대로 이루어지지 않으면서 prediction map이 어느 정도 GT를 잘 따라가고 있음.

## Ablations

![image](https://user-images.githubusercontent.com/52986798/234007383-ae6e6221-bbcf-4eeb-93d0-40f43d6e4ce5.png)

query embedding이 없는 경우에는 성능이 매우 낮음 (query embedding 자체의 중요성 입증). 1-layer query embedding을 사용하면 embedding 사용하지 않았을 경우 대비 성능이 매우 좋아짐. 여기서 Layer-wise query embedding을 활용하면 성능이 더 향상됨. 논문에서는 요소에 따른 성능 향상을 부각시키기 위해 NMA(Neighbor Masked Attention)와 FJ (Feature Jittering)을 1-layer query embedding만 추가한 모델과의 전후 차이를 비교함 (Appendix에서 layer-wise query embedding에 나머지 요소 추가할 때는 성능 차이가 별로 없었음). 역시 성능이 유의하게 증가함을 알 수 있음.

아래의 그림과 같이 layer-wise query decoder에 대한 design choice도 실험을 진행함.

![image](https://user-images.githubusercontent.com/52986798/234008220-82baa482-8e2b-494f-b17e-16749557d9b9.png)

# Conclusion
## Strengths
anomaly detection 시 클래스 별로 학습을 시키는 방식의 비효율성을 지적하고, 이에 multi-class anomaly detection이라는 태스크를 새롭게 제안함.
해당 태스크를 풀 때에 기존 reconstruction-based method가 당면하는 문제점을 설명하고 이를 적절히 해결하고, 그 해결과정이 논리적임.
실험 결과가 매우 강력함.

## Weaknesses
reconstruction based method가 아닌 다른 anomaly detection 방법론들은 identical shortcut 문제를 겪지 않을 수 있음. 예를 들면, patchcore와 같은 memory bank based 방법론들은 학습이 이루어지지 않기 때문에 본질적으로 identical shortcut 문제에서 자유로움. 또한, representation만 받쳐준다면 memory bank 내에서 클래스 별로 충분히 구분이 될 것이므로 multi-class로의 확장이 매우 쉬울 것임. 이러한 부분은 고려하지 않은 채 실험이 되지 않아 아쉬움.
